raft/tracker/tracker.go

```diff
@@ -21,12 +21,27 @@ import (
 	"go.etcd.io/etcd/raft/quorum"
 )
 
+// Config reflects the configuration tracked in a ProgressTracker.
+type Config struct {
+	Voters   quorum.JointConfig
+	Learners map[uint64]struct{}
```

Comment:
This is a little surprising to me. I would have expected Learners to live in `MajorityConfig`. How would a joint configuration where three learners are all atomically upgraded to voting replicas be represented in this model?

Reply:
> How would a joint configuration where three learners are all atomically upgraded to voting replicas be represented in this model?

You start with `voters=(1) learners=(2 3)` and transition to `voters=(1)&(1 2 3) learners=()` and then transition to `voters=(1 2 3) learners=()`

It gets more interesting when you add learners by demoting a voter because we don't at any given point want a peer to be tracked as both a voter and a learner, but naively you need this because you go from

`voters=(1 2 3) learners=()` to `voters=(1 2 3)&(1 2) learners=(3)` so (and this isn't in this PR yet because it's not needed) you get another map `NextLearners` which stashes the learners that will be added when transitioning *out* of the joint config rather than in. I.e you really do

`voters=(1 2 3) learners=()` to `voters=(1 2 3) & (1 2) learners=() next_learners=(3)` and then to `voters=(1 2) learners=(3)`

Basically whenever you see that you need a new learner but in the joint config it's also a voter it goes to next_learners, otherwise to learners. It's more efficient to use learners directly (instead of unconditionally pushing into next_learners) so the peer gets caught up earlier.

> I would have expected Learners to live in MajorityConfig

This could've gone either way, but so far the `quorum` package is really only concerned with quorum computations at the moment. Learners don't influence the quorum.
I don't know if this is relevant to your question, but the ConfState will have to contain nodes, joint nodes, learners, and next learners, i.e. corresponds to a `Config`.

Reply:
> You start with voters=(1) learners=(2 3) and transition to voters=(1)&(1 2 3) learners=() and then transition to voters=(1 2 3) learners=()

But right now the second state would necessarily be `voters=(1)&(1 2 3) learners=(2 3)`

> so (and this isn't in this PR yet because it's not needed) you get another map NextLearners which stashes the learners that will be added when transitioning out of the joint config rather than in

Yeah, so I guess this is what I'm getting at. You'd want to associate a learners map with each side of the joint config. I'm ok with this living outside of the `MajorityConfig`, but we should make it clear that `Learners` only refers to the left side of the joint config.

On that note, would we want to represent this like:
```
type Config struct {
	Voters   quorum.JointConfig
	Learners [2]map[uint64]struct{}
}
```
to mirror the structure in `quorum.JointConfig`?

Reply:
> You'd want to associate a learners map with each side of the joint config.

Sort of, but it's much less symmetric than the joint config (a learner added in the next config may end up in Learners instead of NextLearners), so I'm more comfortable keeping this more explicit. I plan to call this `Learners` and `NextLearners` (added roughly the comment above into the code to make the plan clear).

Reply:
> But right now the second state would necessarily be voters=(1)&(1 2 3) learners=(2 3)

Yep, but also we don't support joint config changes yet :-) I agree that this is awkward but the in-code comment now explains exactly how this will work (and it'll also be the next PR).